python/dgl/distributed/graph_partition_book.py

```diff
@@ -408,101 +494,238 @@ class RangePartitionBook(GraphPartitionBook):
+        Each row represents the Id range of a partition, the first element is the starting Id
+        and the second element is the ending Id.
+    ntypes : a dict
+        map ntype strings to ntype Ids.
+    etypes : a dict
+        map etype strings to etype Ids.
     """
-    def __init__(self, part_id, num_parts, node_map, edge_map):
+    def __init__(self, part_id, num_parts, node_map, edge_map, ntypes, etypes):
         assert part_id >= 0, 'part_id cannot be a negative number.'
         assert num_parts > 0, 'num_parts must be greater than zero.'
         self._partid = part_id
         self._num_partitions = num_parts
-        if not isinstance(node_map, np.ndarray):
-            node_map = F.asnumpy(node_map)
-        if not isinstance(edge_map, np.ndarray):
-            edge_map = F.asnumpy(edge_map)
-        self._node_map = node_map
-        self._edge_map = edge_map
+        self._ntypes = [None] * len(ntypes)
+        self._etypes = [None] * len(etypes)
+        for ntype in ntypes:
+            ntype_id = ntypes[ntype]
+            self._ntypes[ntype_id] = ntype
+        assert all([ntype is not None for ntype in self._ntypes]), \
+                "The node types have invalid Ids."
+        for etype in etypes:
+            etype_id = etypes[etype]
+            self._etypes[etype_id] = etype
+        assert all([etype is not None for etype in self._etypes]), \
+                "The edge types have invalid Ids."
+
+        # This stores the node Id ranges for each node type in each partition.
+        # The key is the node type, the value is a NumPy matrix with two columns, in which
+        # each row indicates the start and the end of the node Id range in a partition.
+        # The node Ids are global node Ids in the homogeneous representation.
+        self._typed_nid_range = {}
+        # This stores the node Id map for per-node-type Ids in each partition.
+        # The key is the node type, the value is a NumPy vector which indicates
+        # the last node Id in a partition.
+        self._typed_node_map = {}
+        max_node_map = np.zeros((num_parts,), dtype=np.int64)
+        for key in node_map:
+            if not isinstance(node_map[key], np.ndarray):
+                node_map[key] = F.asnumpy(node_map[key])
+            assert node_map[key].shape == (num_parts, 2)
+            self._typed_nid_range[key] = node_map[key].reshape(-1, 2)
+            # This is used for per-node-type lookup.
+            self._typed_node_map[key] = np.cumsum(self._typed_nid_range[key][:, 1]
+                                                  - self._typed_nid_range[key][:, 0])
+            # This is used for homogeneous node Id lookup.
+            max_node_map = np.maximum(self._typed_nid_range[key][:, 1], max_node_map)
+        # This is a vector that indicates the last node Id in each partition.
+        # The Id is the global Id in the homogeneous representation.
+        self._node_map = node_map = max_node_map
```

Comment:
Same. Everything is `xxx_map`. The name is not informative.

Reply:
Shadowing the input `node_map` variable here is really bad. It makes the program quite hard to read. Why not directly use `max_node_map`? Also, if `self._node_map` and `self._type_nid_range` really represent the same thing, why save two members?

Reply:
fixed for overwriting `node_map`.
`_node_map` and `_type_nid_range` are diferent. `_node_map` only stores the ending IDs. This is used by `np.searchsorted` for range search. otherwise, we have write a customized version of the range search.